File: apps/mobile/modules/map-memory-manager/android/src/main/java/expo/modules/mapmemorymanager/MapMemoryManagerModule.kt
package expo.modules.mapmemorymanager

import android.view.View
import android.view.ViewGroup
import com.mapbox.maps.MapView
import com.mapbox.maps.TileCacheBudget
import com.mapbox.maps.TileCacheBudgetInMegabytes
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MapMemoryManagerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MapMemoryManager")

    // インメモリタイルキャッシュの上限をMB単位で設定
    Function("setTileCacheBudget") { megabytes: Int ->
      val activity = appContext.currentActivity ?: return@Function
      activity.runOnUiThread {
        findAllMapViews(activity.window.decorView).forEach { mapView ->
          mapView.mapboxMap.setTileCacheBudget(
            TileCacheBudget.valueOf(TileCacheBudgetInMegabytes(megabytes.toLong()))
          )
        }
      }
    }

    // タイルプリフェッチのズームデルタを設定
    Function("setPrefetchZoomDelta") { delta: Int ->
      val activity = appContext.currentActivity ?: return@Function
      activity.runOnUiThread {
        findAllMapViews(activity.window.decorView).forEach { mapView ->
          mapView.mapboxMap.setPrefetchZoomDelta(delta.toByte())
        }
      }
    }

    // 非必須のキャッシュリソースをメモリから解放
    Function("reduceMemoryUse") {
      val activity = appContext.currentActivity ?: return@Function
      activity.runOnUiThread {
        findAllMapViews(activity.window.decorView).forEach { mapView ->
          mapView.mapboxMap.reduceMemoryUse()
        }
      }
    }
  }

  companion object {
    /** ビュー階層を再帰的に走査してMapViewを探す */
    private fun findAllMapViews(view: View): List<MapView> {
      val results = mutableListOf<MapView>()
      findMapViews(view, results)
      return results
    }

    private fun findMapViews(view: View, results: MutableList<MapView>) {
      if (view is MapView) {
        results.add(view)
        return // MapViewの子にMapViewはないので打ち切り
      }
      if (view is ViewGroup) {
        for (i in 0 until view.childCount) {
          findMapViews(view.getChildAt(i), results)
        }
      }
    }
  }
}

File: apps/mobile/modules/map-memory-manager/ios/MapMemoryManagerModule.swift
import ExpoModulesCore
import MapboxMaps

/// Mapbox MapView のメモリ管理 Expo Module
///
/// @rnmapbox/maps が公開していない以下のAPIをブリッジする:
/// - setTileCacheBudget: インメモリタイルキャッシュの上限設定（LRU削除）
/// - setPrefetchZoomDelta: タイルプリフェッチの範囲設定
/// - reduceMemoryUse: 非必須キャッシュリソースの解放
public class MapMemoryManagerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MapMemoryManager")

    /// インメモリタイルキャッシュの上限をMB単位で設定
    Function("setTileCacheBudget") { (megabytes: Int) in
      DispatchQueue.main.async {
        for mapView in Self.findAllMapViews() {
          mapView.mapboxMap.setTileCacheBudget(size: .megabytes(megabytes))
        }
      }
    }

    /// タイルプリフェッチのズームデルタを設定
    Function("setPrefetchZoomDelta") { (delta: Int) in
      DispatchQueue.main.async {
        for mapView in Self.findAllMapViews() {
          mapView.mapboxMap.prefetchZoomDelta = UInt8(delta)
        }
      }
    }

    /// 非必須のキャッシュリソースをメモリから解放
    Function("reduceMemoryUse") {
      DispatchQueue.main.async {
        for mapView in Self.findAllMapViews() {
          mapView.mapboxMap.reduceMemoryUse()
        }
      }
    }
  }

  // MARK: - Private

  /// アプリ内の全MapViewインスタンスをビュー階層から検索
  private static func findAllMapViews() -> [MapView] {
    var results: [MapView] = []
    for scene in UIApplication.shared.connectedScenes {
      if let windowScene = scene as? UIWindowScene {
        for window in windowScene.windows {
          findMapViews(in: window, results: &results)
        }
      }
    }
    return results
  }

  /// 再帰的にビュー階層を走査してMapViewを探す
  private static func findMapViews(in view: UIView, results: inout [MapView]) {
    if let mapView = view as? MapView {
      results.append(mapView)
      return // MapViewの子にMapViewはないので打ち切り
    }
    for subview in view.subviews {
      findMapViews(in: subview, results: &results)
    }
  }
}
